### PR TITLE
build: make rollup generate a depfile for the bundle build

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -122,15 +125,85 @@ function resolveGenerated() {
...
   };
 }
 
+function generateDepFile({ outputFile, sourceFiles = [], configFiles = [] }) {
+  let timestamp = Date.now();
+
+  // Save the depfile just before the node process exits.
+  process.once("beforeExit", () =>
+    writeDepFile({ outputFile, sourceFiles, configFiles, timestamp })
+  );
```
Comment (ry) [inline rollup.config.js]: Is this the only hook to do this sort of thing? What if there's an error in rollup?

File: rollup.config.js [hunk 2]
```diff
@@ -122,15 +125,85 @@ function resolveGenerated() {
...
+
+  return {
+    name: "depfile",
+    load(sourceFile) {
+      // The 'globals' plugin adds generated files that don't exist on disk.
+      // Don't add them to the depfile.
+      if (/^[0-9a-f]{30}$/.test(sourceFile)) {
```
Comment (ry) [inline rollup.config.js]: ok?  odd.

File: rollup.config.js [hunk 3]
```diff
@@ -122,15 +125,85 @@ function resolveGenerated() {
...
+    }
+  };
+}
+
+function writeDepFile({ outputFile, sourceFiles, configFiles, timestamp }) {
+  const buildDir = process.cwd();
+  const outputDir = path.dirname(outputFile);
+
+  // Assert that the discovered bundle inputs are files that exist on disk.
+  sourceFiles.forEach(f => fs.accessSync(f));
```
Comment (ry) [inline rollup.config.js]: This throws if the file doesn't exist?  I feel like `assert(fs.existsSync(f))` would be more readable.

File: rollup.config.js [hunk 4]
```diff
@@ -122,15 +125,85 @@ function resolveGenerated() {
...
+  fs.utimesSync(depFile, timestamp, timestamp);
+
+  // Renders path to make it suitable for a depfile.
+  function formatPath(filename) {
+    // Make the path relative to the root build directory.
+    filename = path.relative(buildDir, filename);
+    // Use forward slashes on Windows.
+    if (process.platform === "win32") {
+      filename = filename.replace(/\\/g, "/");
+    }
```
Comment (ry) [inline rollup.config.js]: This won't be necessary in deno, once #957 is done

File: rollup.config.js (hunk 1 above)
Comment (piscisaureus) [reply rollup.config.js]: 🤷‍♂️  ... node will crash before the depfile gets written I suppose?

File: rollup.config.js (hunk 3 above)
Comment (piscisaureus) [reply rollup.config.js]: Oh I was so entertained that I finally found a place where fs.access() had some modest utility.

**Pull request merged**